File: src/go/partition/l/par.go
package large

import (
	"runtime"
	"sync"
	"sync/atomic"
)

func large_sum_par(arr []int, size int, index int64) int {
	var sum int = 0

	for i := 0; i < size; i++ {
		if index&(1<<i) != 0 {
			sum += arr[i]
		}
	}

	return sum
}

func Par(arr []int) int32 {
	var result int32 = 0

	var combs int64 = 1 << (len(arr) - 1)
	var all int64 = (1 << len(arr)) - 1

	// Calculate the total sum of the array
	var problem_sum int = large_sum_par(arr, len(arr), all)
	if problem_sum%2 != 0 {
		return 0
	}
	var half_sum = problem_sum / 2

	// Shared variable to indicate if a solution is found
	// var found int32 = 0
	var wg sync.WaitGroup

	threads := runtime.NumCPU()
	minProblems := combs / int64(threads)

	// Parallelize the computation
	for w := 0; w < threads; w++ {
		wg.Add(1)

		go func(w int) {
			defer wg.Done()

			for i := 0; i <= int(minProblems); i++ {
				var index int64 = int64(i*threads + w)
				if index >= combs {
					break
				}

				if atomic.LoadInt32(&result) == 1 {
					break
				}

				if large_sum_par(arr, len(arr), index) == half_sum {
					atomic.StoreInt32(&result, 1)
				}
			}

		}(int(w))
	}

	// Wait for all goroutines to complete
	wg.Wait()

	return result
}

File: src/julia/partition/l/par.jl
module Par

using Base.Threads
export large_par

function large_sum_par(arr::Vector{Int64}, size::Int, index::Int64)
    sum = 0

    for i in 1:size
        if index & (1 << (i - 1)) != 0
            sum += arr[i]
        end
    end

    return sum
end # large_sum_par

function large_par(arr::Vector{Int64})
    size = length(arr)
    combs = 1 << (size - 1)
    all = (1 << size) - 1

    problem_sum = large_sum_par(arr, size, all)
    if problem_sum % 2 != 0
        return [false]
    end
    half_sum = problem_sum ÷ 2

    found = Threads.Atomic{Bool}(false)

    Threads.@threads for i in 1:combs
        if found[]
            continue
        end

        sum = large_sum_par(arr, size, i)
        if sum == half_sum
            found[] = true
        end
    end

    return [found[]]
end # large_par

end # module
